corgi/core/models.py

```diff
@@ -1986,6 +1999,54 @@ def get_upstreams_purls(self, using: str = "read_only") -> set[str]:
 
+    def disassociate_with_product(self, product_ref: ProductModel):
+        """Disassociate this component with the passed in ProductModel and any child ProductModels
+         in that product's hierarchy. This is the reverse of what happens in save_product_taxonomy
+        and is used when a product is removed from product_definitions, or moved to a new place
+        in the product hierarchy, which also requires re-processing those products from scratch
+        Removing proudctvariants is a no-op where a stream does not have errata_info. It's easier to
+        just remove productvariants everytime then check for errata_info in a stream.
+        """
+        if isinstance(product_ref, Product):
+            self.productvariants.remove(*product_ref.productvariants.get_queryset())
+            self.productstreams.remove(*product_ref.productstreams.get_queryset())
+            self.productversions.remove(*product_ref.productversions.get_queryset())
+            self.products.remove(product_ref)
+        elif isinstance(product_ref, ProductVersion):
+            self.productvariants.remove(*product_ref.productvariants.get_queryset())
+            self.productstreams.remove(*product_ref.productstreams.get_queryset())
+            self.productversions.remove(product_ref)
+            self._check_and_remove_orphaned_product_refs(product_ref, "Product")
+        elif isinstance(product_ref, ProductStream):
+            self.productvariants.remove(*product_ref.productvariants.get_queryset())
+            self.productstreams.remove(product_ref)
+            self._check_and_remove_orphaned_product_refs(product_ref, "ProductVersion")
+            self._check_and_remove_orphaned_product_refs(product_ref, "Product")
+        elif isinstance(product_ref, ProductVariant):
+            self.productvariants.remove(product_ref.productvariants)
+            self._check_and_remove_orphaned_product_refs(product_ref, "ProductStream")
+            self._check_and_remove_orphaned_product_refs(product_ref, "ProductVersion")
+            self._check_and_remove_orphaned_product_refs(product_ref, "Product")
+        else:
+            raise NotImplementedError(
+                f"Add class {type(product_ref)} to Component.disassociate_with_product()"
+            )
+
+    def _check_and_remove_orphaned_product_refs(self, product_ref, product_model_name):
+        """Remove product_models from this component where there are no remaining product children
+        or the remaining product children don't share this product_model as a parent."""
+        product_node_level = MODEL_NODE_LEVEL_MAPPING[product_model_name]
+        product_attribute = NODE_LEVEL_ATTRIBUTE_MAPPING[product_node_level]
+        query = getattr(self, NODE_LEVEL_ATTRIBUTE_MAPPING[product_node_level + 1]).get_queryset()
+        remaining_product_child_count = query.count()
+        parents_of_remaining_children = query.values_list(product_attribute, flat=True)
```

Comment:
You can avoid making an extra query to the DB if we don't need `.distinct()` here:
```suggestion
        parents_of_remaining_children = child_products_query.values_list(product_attribute, flat=True)
        remaining_product_child_count = len(parents_of_remaining_children)
```
Or do we want a distinct() list of the parents?

Reply:
Good idea. Probably want distinct actually, to avoid removing the same parent more than once.